[region-isolation] Implement function sub typing rules

gottesmm opened the pull request:
Just taking care of this real quick.

rdar://127675288

gottesmm committed:
[sending] Do not let protocol function reqs with a sending result be witnessed by a function without a sending result.

This is important since the witness is not promising to return a disconnected
value and the caller of the requirement is going to accept that.

rdar://127675288

gottesmm committed:
[sending] Implement non-protocol function subtyping rules.

rdar://127675288

gottesmm commented:
@swift-ci smoke test

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -3235,6 +3235,12 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
+  // () -> sending T can be a subtype of () -> T... but not vis-a-versa.
+  if (func1->hasSendingResult() != func2->hasSendingResult() &&
+      (!func1->hasSendingResult() || kind < ConstraintKind::Subtype)) {
+    return getTypeMatchFailure(locator);
```
I think both of these new checks require a tailored fix to explain what exactly went wrong because otherwise users would be left to figure out why two function types didn't match which might not always be obvious for complex types... Doesn't have to be implemented in this PR though.

gottesmm committed:
[sending] Check for a nullptr in AssociatedTypeInference when calling computeFailureTypeWitness.

The problem happens when a type conforms to AsyncIterator.next when next in the
protocol returns its value as sending, but the witness does not have sending.
This results in the function subtyping rules rejecting the witness... but we
still leave in the witness as a nullptr... so we need to handle it here.

rdar://129300953

gottesmm committed:
Simplify test to only use target-typecheck-verify-swift since we are testing typecheck errors.

gottesmm committed:
[sending] Fix a few bugs around closure inference of sending parameters and results.

I found this while writing tests for the earlier part of this work. Since this
is also type checking work, I am just folding this work into that work.

gottesmm committed:
[sending] Fix type inference for sending results of closures.

The previous commit fixed things like:

```swift
let x: () -> sending String = { "" }
```

This commit fixes this test case:

```swift
let x = { () -> sending String in "" }
```

gottesmm commented:
@swift-ci smoke test

gottesmm committed:
Fix a test

gottesmm commented:
@swift-ci smoke test

gottesmm committed:
[sending] Improve the sending mismatch errors and make them warnings when not in swift 6.

This will ensure that we do not break anyone who has adopted APIs like
CheckedContinuation.resume that now have sending parameters.

An example of where this can come up is shown by the ProcessType in SwiftToolsCore:

```swift
@available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
@discardableResult
public func waitUntilExit() async throws -> ProcessResult {
    try await withCheckedThrowingContinuation { continuation in
        DispatchQueue.processConcurrent.async {
            self.waitUntilExit(continuation.resume(with:))
        }
    }
}
```

This fails to compile since self.waitUntilExit doesn't expect a function that
takes a sending parameter. We want to give people time to fix such issues.

gottesmm commented:
@swift-ci smoke test

ktoso commented on test/Concurrency/transfernonsendable_functionsubtyping.swift:
```diff
@@ -0,0 +1,111 @@
+class NonSendableKlass {}
+
+protocol ProtocolWithSendingReqs {
+  func sendingResult() -> sending NonSendableKlass // expected-note {{}}
```
what is expected note {{}}  -- "any note"? First time seeing this tbh; worth spelling it out?

gottesmm replied:
The reason I did this is that the message here isn't important to what I am testing but it is still a diagnostic.

ktoso commented on include/swift/Sema/CSFix.h:
```diff
@@ -474,6 +474,17 @@ enum class FixKind : uint8_t {
+  /// parameter.
+  ///
+  /// 2. Where we have a function that expects a function typed parameter with a
+  /// sending result, but is passed a function typeed parameter without a
```
minor typo:

```suggestion
  /// sending result, but is passed a function typed parameter without a
```

ktoso reviewed (approved):
Looks OK, I'm not 100% sure about the rule but it sounds good from thinking about it for a while... LGTM

gottesmm commented:
macOS failure is something unrelated in lldb.

gottesmm commented:
@swift-ci smoke test macOS platform

gottesmm committed:
[sending] Make the protocol errors only occur in swift 6.

For now, just turn off the error in swift 5 mode. I wanted to make this an error
but right now I do not have all of the time to do it so this is good in the
short term so that people can adopt sending on protocols without breaking people
using their framework in executables still compiling in swift 5.

gottesmm commented:
@swift-ci smoke test

xedin commented on include/swift/Sema/CSFix.h:
```diff
@@ -2648,6 +2659,49 @@ class TreatEphemeralAsNonEphemeral final : public AllowArgumentMismatch {
+/// that this is being sent to does not enforce that invariants within its body.
+class AllowSendingMismatch final : public ContextualMismatch {
+public:
+  enum class Kind {
```
I don't think this is necessary, it could be determined based on the locator associated wit the fix.

xedin commented on include/swift/Sema/CSFix.h:
```diff
@@ -2648,6 +2659,49 @@ class TreatEphemeralAsNonEphemeral final : public AllowArgumentMismatch {
+
+  static AllowSendingMismatch *create(ConstraintSystem &cs,
+                                      ConstraintLocator *locator, Type srcType,
+                                      Type dstType, Kind kind);
```
Nit: Based on all other `create` methods let's use - `cs, src, dst, locator` ordering.

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -2306,6 +2306,28 @@ class NonEphemeralConversionFailure final : public ArgumentMismatchFailure {
+  bool diagnoseAsError() override;
+};
+
+class SendingOnFunctionResultMismatchFailure final : public ContextualFailure {
```
It looks like this split is not necessary, the diagnostic is anchored on a simplified locator so it would point to the right spot.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -8020,6 +8020,24 @@ bool NonEphemeralConversionFailure::diagnoseAsError() {
+}
+
+bool SendingOnFunctionResultMismatchFailure::diagnoseAsError() {
+  emitDiagnosticAt(getLoc(), diag::sending_function_wrong_sending,
```
Why not just `emitDiagnostic` since you are using `getLoc()` anyway?

xedin commented on lib/Sema/CSFix.cpp:
```diff
@@ -1884,6 +1884,34 @@ std::string TreatEphemeralAsNonEphemeral::getName() const {
+        solution, getFromType(), getToType(), getLocator(), fixBehavior);
+    return failure.diagnose(asNote);
+  }
+  case Kind::Result: {
```
It doesn't look like this split brings much to the table, the difference could be handled internally to the diagnostic.

ahoppen commented on lib/Parse/ParseType.cpp:
```diff
@@ -1555,6 +1555,8 @@ bool Parser::canParseType() {
     consumeToken();
   } else if (Tok.isContextualKeyword("each")) {
     consumeToken();
+  } else if (Tok.isContextualKeyword("sending")) {
+    consumeToken();
```
Is this correct? My impression is that only parameters can be annotated as `sending` and not arbitrary types. Looking at this change, I think it might cause the following to stop compiling because `sending` now gets consumed as a contextual keyword instead of as the type name.

```swift
struct sending {}
let a: sending = sending()
func test(x: sending) {}
```

gottesmm replied:
@ahoppen I just tested this locally and it isn't an issue.

ahoppen replied:
I just attached a debugger to it and understand why my example didn’t hit the issue that I was thinking about but this one does

```swift
struct sending {}
func test() {
  _ = { () -> sending in
  }
}
```

I think the proper solution here is to check if `sending` is followed by something that satisfies `canParseType`. If not, we probably need to backtrack and check if `sending` itself can be used as the type name.
